129 Create background thread to check agent's critical processes

bodiakof opened the pull request:
Command to run critical_processes_check.py:

MONITOR_INTERVAL=10 python -m agents.dev.critical_processes_check _agent's_name_

(MONITOR_INTERVAL= is optional, to specify needed interval)

bodiakof committed:
feat: add standalone process monitor script for agent with background thread and logging

- Run process checks in a dedicated background thread using threading and stop events.
- Log process states (RUNNING, DOWN, RECOVERED) to both file and console using logconfig.ini.

bodiakof committed:
commit to switch to develop :)

bodiakof committed:
Merge branch 'develop' into 129-create_background_thread_to_check_critical_processes

bodiakof committed:
Merge branch 'develop' into 129-create_background_thread_to_check_critical_processes

bodiakof committed:
Modify criical_processes_check.py to suport is_service_healthy and maybe_restart_service methods

mehalyna commented on agents/dev/critical_processes_check.py:
```diff
@@ -0,0 +1,117 @@
+logging.config.fileConfig(
+    conf_path,
+    defaults={
+        'agent_name': 'ntp',
```
If multiple agents might reuse this script structure, consider retrieving the agent name dynamically (e.g., from the agent object) to make the config more reusable.

mehalyna commented on agents/dev/critical_processes_check.py:
```diff
@@ -0,0 +1,117 @@
+        args=(agent, 10, stop_event),
+        name='ntp-monitor'
+    )
+    t.setDaemon(True)
```
Prefer using the `daemon=True` argument in the Thread constructor:
```
t = threading.Thread(
    target=monitor_processes,
    args=(agent, 10, stop_event),
    name='ntp-monitor',
    daemon=True
)
```

bodiakof replied:
Python 2.6’s Thread doesn’t accept `daemon=` in its constructor and `TypeError: __init__() got an unexpected keyword argument 'daemon'` gets thrown.

mehalyna commented on agents/dev/critical_processes_check.py:
```diff
@@ -0,0 +1,117 @@
+
+        # Check each critical process
+        for proc in agent.critical_processes:
+            agent.processes = [proc]
```
Overwriting `agent.processes` inside the loop can be risky if the agent uses this attribute elsewhere. Consider refactoring `_is_process_running(proc)` to accept the process name directly without modifying internal state.

mehalyna commented on agents/dev/critical_processes_check.py:
```diff
@@ -0,0 +1,117 @@
+    stop_event = threading.Event()
+    t = threading.Thread(
+        target=monitor_processes,
+        args=(agent, 10, stop_event),
```
Allow the interval to be configurable via CLI argument or environment variable for better control in production.

mehalyna commented on agents/dev/critical_processes_check.py:
```diff
@@ -0,0 +1,117 @@
+        logger.info('Interrupted, stopping monitor...')
+        stop_event.set()
+        t.join(5)
+        sys.exit(0)
```
Add a final log line before exiting to indicate completion:
```
logger.info('Monitor shutdown complete.')
```

mehalyna commented on agents/ntp/ntp.py:
```diff
@@ -36,27 +36,48 @@ def service_healthy(
+        # treat as down.
+        ntp_running = False
+        for proc in self.critical_processes:
+            self.processes = [proc]
```
This alters `self.processes`, which might affect future checks or logic. To prevent unintended side effects, consider modifying `_is_process_running()` to accept a process name directly, e.g.:
```
def _is_process_running(self, proc_name=None):
    ...
```
And then use:
```
if self._is_process_running(proc):
```

mehalyna commented on agents/ntp/ntp.py:
```diff
@@ -36,27 +36,48 @@ def service_healthy(
-        return True
+        )
+
+        return False
```
Always returns `False` after trying restarts. After attempting restarts, consider checking again whether services recovered, and return `True` if they did:
```
# After restarts, re-check
if self.is_ssh_service_active() and any(
    self._is_process_running(proc) for proc in self.critical_processes
):
    return True
return False
```
This adds reliability for monitoring systems.

mehalyna commented on agents/ntp/ntp.py:
```diff
@@ -36,27 +36,48 @@ def service_healthy(
+            restart_service(self.logger, self.fallback_logger, service)
 
         maybe_log_message(
-            'All services are heathy and running',
+            'Finished attempts to restart services: %s' % ', '
+            .join(inactive_services),
```
Move join to a variable first for clarity and logging reuse:
```
services_str = ', '.join(inactive_services)
maybe_log_message(
    f'Finished attempts to restart services: {services_str}',
    ...
)
```

bodiakof committed:
Refactor ServerAgent._is_process_running to accept a single process name, avoiding in-place mutation of self.processes

bodiakof committed:
Refactor and generalize critical_processes_check.py

- Replace NTP-specific hard-coded script with a universal critical_processes_check.py CLI utility
- Dynamically import any ServerAgent subclass (ntp, dns, or smtp) based on a positional argument
- Introduce --interval (or MONITOR_INTERVAL env var) to control the polling frequency
- Remove all inline logging.config.fileConfig calls; instead call the agent’s own setup_logging() for consistent file+console handlers
- Implement a console fallback handler if, for any reason, no handlers were configured by setup_logging()

bodiakof committed:
Refactor NTPAgent maybe_restart_service for clarity, safety, and robustness

- Refactored maybe_restart_service to avoid mutating self.processes during process checks.
  Now uses the updated _is_process_running(proc_name=...) method directly.
- After service restart attempts, health is rechecked for both SSH and NTP services to confirm recovery.
  The method now returns True if services successfully recover post-restart.
- Improved log clarity:
  - Moved service list string creation (', '.join(...)) to a separate variable for reuse.
  - Log messages now clearly indicate whether services recovered or remain down after restarts.
- Updated method and class-level docstrings for accuracy and completeness.

NovitskaMariia commented on agents/dev/critical_processes_check.py:
```diff
@@ -0,0 +1,168 @@
+    )
+    parser.add_argument(
+        'agent_type',
+        choices=['ntp', 'dns', 'smtp'],
```
Add 'web' to choices

YelyzavetaV commented on agents/agent.py:
```diff
@@ -442,11 +454,13 @@ def _is_process_running(self):
+                    if re.search(pattern, line):
+                        return True
+
+            return False
```
Agent is currently being refactored. Let's revert it to the way it is in `develop` for consistency.

YelyzavetaV commented on agents/dev/critical_processes_check.py:
```diff
@@ -0,0 +1,168 @@
+    return getattr(module, class_name)
+
+
+def monitor_processes(agent, interval, stop_event):
```
`is_service_healthy` is to wrap a dedicated method, which checks all critical processes. 
`monitor_processes` worker should do either one or the other: check health status or check processes.

bodiakof committed:
Add 'web' to parser's choices

bodiakof committed:
Revert changes in _is_process_running back

bodiakof committed:
Modify monitor_processes to check only processes

bodiakof committed:
Merge branch 'develop' into 129-create_background_thread_to_check_critical_processes
